### PR TITLE
OpenStack Heat provider (Step 1 of 2)

### PR DESCRIPTION
**Pull request by ohtake**
#21737 is divided into 2 pull requests to make CLA bot work.

Reviews in #21737 have been applied to this pull request.
- [x] Change directory name from `openstack` to `openstack-heat`.
- [x] Use `host-gw` flannel for better performance.

### CODE REVIEW

File: cluster/openstack-heat/kubernetes-heat/fragments/run-salt.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+set -o nounset
+set -o pipefail
+
+. /etc/sysconfig/heat-params
+
+rm -rf /kube-install
+mkdir -p /kube-install
+cd /kube-install
+
+curl "$KUBERNETES_SERVER_URL" -o kubernetes-server.tar.gz
```
Comment (brendandburns) [inline cluster/openstack-heat/kubernetes-heat/fragments/run-salt.sh]: `${KUBERNETES_SERVER_URL}`

File: cluster/openstack-heat/kubernetes-heat/fragments/run-salt.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+curl "$KUBERNETES_SALT_URL" -o kubernetes-salt.tar.gz
```
Comment (brendandburns) [inline cluster/openstack-heat/kubernetes-heat/fragments/run-salt.sh]: Here too.

File: cluster/openstack-heat/kubernetes-heat/fragments/write-heat-params.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#cloud-config
+merge_how: dict(recurse_array)+list(append)
+write_files:
+  - path: /etc/sysconfig/heat-params
+    owner: "root:root"
+    permissions: "0644"
+    content: |
+      KUBERNETES_SERVER_URL="$KUBERNETES_SERVER_URL"
```
Comment (brendandburns) [inline cluster/openstack-heat/kubernetes-heat/fragments/write-heat-params.yaml]: `${...}` for env vars.

**Comment by brendandburns** [comment]
Minor stuff in the scripts.  I'm assuming the heat YAML works, I have no idea.

@k8s-bot ok to test.

This looks fine to me, but sending over to @bgrant0607 for final LGTM since he was assigned #21737

File: cluster/openstack-heat/kubernetes-heat/fragments/run-salt.sh (hunk 2 above)
Comment (ohtake) [inline cluster/openstack-heat/kubernetes-heat/fragments/run-salt.sh]: Done. I have added braces for environment variables.

File: cluster/openstack-heat/kubernetes-heat/fragments/write-heat-params.yaml (hunk 3 above)
Comment (ohtake) [inline cluster/openstack-heat/kubernetes-heat/fragments/write-heat-params.yaml]: `$KUBERNETES_SERVER_URL` is not an environment variable but a heat `str_replace` param name.
Do you think it is better to change heat `str_replace` param names too?

**Comment by quinton-hoole** [comment]
Has anyone run the e2e conformance tests against this yet?

https://github.com/kubernetes/kubernetes/blob/c6e995a824094a96f7d43a25e897283f83a12997/docs/devel/e2e-tests.md . See the section on conformance tests.

**Comment by ohtake** [comment]
The issue of multiple authors in a PR will be solved.
Changes in this PR has been added to #21737.

I will close this PR.

**Pull request closed without merging**